File: 백준/Gold/15684. 사다리 조작/사다리 조작.py
import sys
input = sys.stdin.readline
def move():
    for i in range(n):
        now = i
        for j in range(h):
            # 가로선 존재
            if board[j][now]:
                now += 1
            elif now>0 and board[j][now-1]:
                now -= 1
        if now != i:
            return False
    return True



def dfs(cnt, x, y):
    global res
    if move():
        res = min(res, cnt)
        return

    elif cnt >= 3 or cnt >= res:
        return

    for i in range(x, h):
        if i == x:
            now = y
        else:
            now = 0

        for j in range(now, n-1):
            if not board[i][j] and not board[i][j+1]:
                if j > 0 and board[i][j-1]:
                    continue
                board[i][j] = 1
                dfs(cnt+1, i, j+2)
                board[i][j] = 0



# n개의 세로선과 m개의 가로선
n, m, h = map(int,input().split())
board = [[0] * n for _ in range(h)]

for _ in range(m):
    a, b = map(int,input().split())
    board[a-1][b-1] = 1

res = float('inf')
dfs(0,0,0)
if res > 4:
    print(-1)
else:
    print(res)

File: 백준/Gold/15684. 사다리 조작/사다리 조작.js
const fs = require("fs");
const inputData = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

// 보드판은 n*h로 생겼고, m개의 가로선이 주어져 있음
const [n, m, h] = inputData[0].split(" ").map(Number);

const board = Array.from({ length: h }, () => Array(n).fill(0));

for (let i = 0; i < m; i++) {
  const [a, b] = inputData[i + 1].split(" ").map(Number);
  board[a - 1][b - 1] = 1;
}

const valiable = [];
let res = 0;
let check = 0;
const now = [];

for (let i = 0; i < h; i++) {
  for (let j = 0; j < n - 1; j++) {
    if (board[i][j] === 0) {
      valiable.push([i, j]);
    }
  }
}

// 내려가는 것을 구현하는 함수
function lad() {
  for (let i = 0; i < n; i++) {
    // i 세로선을 시작으로 내려가기 !!!@!@
    nowRow = i;
    for (let nowCol = 0; nowCol < h; nowCol++) {
      if (board[nowCol][nowRow] === 1) {
        nowRow += 1;
      }
      // 왼쪽 확인
      else if (nowRow > 0 && board[nowCol][nowRow - 1] === 1) {
        nowRow -= 1;
      }
    }
    // i의 결과가 i가 아님 -> 답이 될 수 없음
    if (nowRow !== i) {
      return false;
    }
  }
  return true;
}

// 가로선 그려주는 함수
function draw(now) {
  for (let i = 0; i < now.length; i++) {
    board[now[i][0]][now[i][1]] = 1;
  }
  if (lad()) {
    check = 1;
    return true;
  }

  for (let i = 0; i < now.length; i++) {
    board[now[i][0]][now[i][1]] = 0;
  }
  return false;
}

function dfs(depth, res, idx) {
  // if (check) {
  //   return;
  // }
  if (depth === res) {
    draw(now);
    return;
  }
  for (let i = idx; i < valiable.length; i++) {
    now.push(valiable[i]);
    dfs(depth + 1, res, i + 1);
    now.pop();
  }
}

while (res < 4) {
  dfs(0, res, 0);
  if (check) {
    console.log(res);
    break;
  }
  res += 1;
}

if (!check) {
  console.log(-1);
}
